infrastructure/smart-contracts/contracts/upgrades/node-license/NodeLicense9.sol

```diff
@@ -0,0 +1,872 @@
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/StringsUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/Base64Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "../../upgrades/referee/Referee10.sol";
+import "../../RefereeCalculations.sol";
+
+interface IAggregatorV3Interface {
+    function latestAnswer() external view returns (int256);
+}
+
+contract NodeLicense9 is ERC721EnumerableUpgradeable, AccessControlUpgradeable  {
+    using StringsUpgradeable for uint256;
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+    CountersUpgradeable.Counter private _tokenIds;
+
+    address payable public fundsReceiver;
+
+    uint256 public maxSupply; // Maximum number of licenses that can be minted
+
+    // Define the pricing table
+    Tier[] private pricingTiers;
+
+    uint256 public referralDiscountPercentage;
+    uint256 public referralRewardPercentage;
+
+    // Boolean to control whether referral rewards can be claimed
+    bool public claimable;
+
+    // Mapping from token ID to minting timestamp
+    mapping (uint256 => uint256) public _mintTimestamps;
+
+    // Mapping from promo code to PromoCode struct
+    mapping (string => PromoCode) private _promoCodes;
+
+    // Mapping from referral address to referral reward
+    mapping (address => uint256) private _referralRewards;
+
+    // Mapping from token ID to average cost, this is used for refunds over multiple tiers
+    mapping (uint256 => uint256) public _averageCost;
+    
+    // Mapping for whitelist to claim NFTs without a price
+    mapping (address => uint16) public whitelistAmounts;
+
+    /**
+     *  @dev New promo code mappings are ONLY used for 
+     *  reward tracking of XAI and esXAI rewards
+     *  The original promo code mapping is used for promo
+     *  code creation, removal, activation and recipient tracking
+     * 
+     *  @notice The new mappings should not be checked for 
+     *  promo code existence or activation, they simply track 
+     *  the referral rewards for XAI and esXAI 
+     */
+
+    // Mapping from promo code to PromoCode struct for Xai
+    mapping (string => PromoCode) private _promoCodesXai;
+
+    // Mapping from promo code to PromoCode struct for esXai
+    mapping (string => PromoCode) private _promoCodesEsXai;
+
+    // Mapping from referral address to referral reward
+    mapping (address => uint256) private _referralRewardsXai;
+
+    // Mapping from referral address to referral reward
+    mapping (address => uint256) private _referralRewardsEsXai;
+
+    // Chainlink Eth/USD price feed
+    IAggregatorV3Interface internal ethPriceFeed;
+
+    //Chainlink XAI/USD price feed
+    IAggregatorV3Interface internal xaiPriceFeed;
```

Comment:
Where is this coming from? `IAggregatorV3Interface`

I don't see it imported or defined anywhere?

Reply:
It was coming from the PoolFactory2 importing NodeLicense8 which is not used, I think its better to remove the import from there, thanks!

Reply:
Understood, looks good!